# LaboratoryTask3


**Jade Natasha S. Ragot**

**DS4A**



**Laboratory Task 3**

*(01_Understanding Deep Learning)*


## **Previously...**

In [1]:
import numpy as np

x = np.array([1, 0, 1])
y = np.array([1])
lr = 0.001

W_hidden = np.array([
    [0.2, -0.3],
    [0.4,  0.1],
    [-0.5, 0.2]
])
theta_hidden = np.array([-0.4, 0.2])

W_output = np.array([-0.3, -0.2])
theta_output = 0.1

def relu(z):
    return np.maximum(0, z)

def relu_derivative(z):
    return (z > 0).astype(float)   # 1 where z>0, else 0

In [2]:
Z_hidden = x @ W_hidden + theta_hidden
H = relu(Z_hidden)
Z_output = H @ W_output + theta_output
y_hat = relu(Z_output)
E = 0.5 * (y - y_hat) ** 2

print("Z_hidden =", Z_hidden)
print("H =", H)
print("Z_output =", Z_output)
print("y_hat =", y_hat)
print("E =", E)

Z_hidden = [-0.7  0.1]
H = [0.  0.1]
Z_output = 0.08
y_hat = 0.08
E = [0.4232]


<p style="font-family:Times New Roman; text-align:justify; font-size:15px">
    <b>Instruction:</b> Perform a forward and backward propagation in python using the inputs from <b>Laboratory Task 2</b>
</p>

```python
x = np.array([1, 0, 1])
y = np.array([1])

# use relu as the activation function.

# learning rate
lr = 0.001
```

In [4]:
# Output layer error (δo)

delta_o = (y_hat - y) * relu_derivative(Z_output)
print("delta_o =", delta_o)

delta_o = [-0.92]


In [5]:
# Hidden layer error (δh)

delta_h = (delta_o * W_output) * relu_derivative(Z_hidden)
print("delta_h =", delta_h)

delta_h = [0.    0.184]


>> **delta_o * W_output** = −0.92 × [−0.3, −0.2] = [0.276, 0.184]

— this is "how much did each hidden neuron contribute to the output error."


This **[0.276, 0.184]** isn't the final δh yet — it's only "half" the formula. Then multiply by

>> **relu_derivative(Z_hidden) = [0, 1] (since Z1=−0.7 was negative, Z2=0.1 was positive)**

That zeroes out H1's contribution entirely — makes sense, since H1 was already "dead" (0) going forward, so it can't receive any error signal going backward either.

In [7]:
grad_W_output = H * delta_o
grad_theta_output = delta_o

grad_W_hidden = np.outer(x, delta_h)
grad_theta_hidden = delta_h

print("grad_W_output =", grad_W_output)
print("grad_theta_output =", grad_theta_output)
print("grad_W_hidden =\n", grad_W_hidden)
print("grad_theta_hidden =", grad_theta_hidden)

grad_W_output = [-0.    -0.092]
grad_theta_output = [-0.92]
grad_W_hidden =
 [[0.    0.184]
 [0.    0.   ]
 [0.    0.184]]
grad_theta_hidden = [0.    0.184]


In [8]:
W_output_new = W_output - lr * grad_W_output
theta_output_new = theta_output - lr * grad_theta_output

W_hidden_new = W_hidden - lr * grad_W_hidden
theta_hidden_new = theta_hidden - lr * grad_theta_hidden

print("Updated W_output =", W_output_new)
print("Updated theta_output =", theta_output_new)
print("Updated W_hidden =\n", W_hidden_new)
print("Updated theta_hidden =", theta_hidden_new)

Updated W_output = [-0.3      -0.199908]
Updated theta_output = [0.10092]
Updated W_hidden =
 [[ 0.2      -0.300184]
 [ 0.4       0.1     ]
 [-0.5       0.199816]]
Updated theta_hidden = [-0.4       0.199816]


We can see how after one round of forward and backward propagation, the network's weights changed only a tiny bit because the learning rate of 0.001 was set very small on purpose, so training happens gradually instead of big jumps.

Also, some weights didn't change at all that's because H1 got "killed" by ReLU during the forward pass (its value was 0), so it had no way to send an error signal backward either so a neuron that outputs 0 stops learning.

In conclusion, one pass isn't enough to fix the error. The network would need to repeat this forward + backward process many times before ŷ actually gets close to the target y = 1.

